HTTPs support for the job client

joncarter1 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This PR enables HTTPS support within the Ray JobSubmissionClient.
e.g. 
```
client = JobSubmissionClient(address, verify="/path/to/cert")
client.submit_job(...)
```

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

Closes #30888

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

I have also tested that `aiohttp` HTTPS connections (tailing job logs) work using a simple nginx HTTPS proxy.

joncarter1 commented:
Removing stale label

joncarter1 commented:
Hey @architkulkarni, as someone who's touched these files recently would you be able to review this?

architkulkarni commented on dashboard/modules/dashboard_sdk.py:
```diff
@@ -221,6 +224,20 @@ def __init__(
+            if os.path.isdir(self._verify):
+                cafile, capath = None, self._verify
+            else:
+                cafile, capath = self._verify, None
```
```suggestion
                cafile, capath = self._verify, None
```
Should we also check if it's an existing file and raise a ValueError with a friendly message if it's neither an existing file nor existing dir?

joncarter1 replied:
Makes sense! I had delegated that to the underlying libraries but no harm in adding that in here.

architkulkarni commented on dashboard/modules/dashboard_sdk.py:
```diff
@@ -221,6 +224,20 @@ def __init__(
+            if self._verify is False:
+                self._ssl_context = False
+            else:
+                self._ssl_context = None
```
I see, it's added below. Not sure why all those existing checks are in JobSubmissionClient but not SubmissionClient, but I guess we can leave it as is for now.

architkulkarni commented on dashboard/modules/job/sdk.py:
```diff
@@ -60,6 +60,8 @@ class JobSubmissionClient(SubmissionClient):
         headers: Headers to use when sending requests to the HTTP job server, used
             for cases like authentication to a remote cluster.
+        verify: Boolean indication to verify the server's TLS certificate or a path to
+            a file or directory of trusted certificates.
```
We should state the default in this docstring.

shrekris-anyscale replied:
+1

shrekris-anyscale replied:
Will clients need to explicitly set `verify` to `False` to opt-out of TLS?

joncarter1 replied:
@shrekris-anyscale that's correct.

`verify=True` is the default for the `requests.request` function so default client behaviour is unchanged.

architkulkarni reviewed (commented):
This looks good to me so far, and the new API makes sense because it matches the API in the `requests` library (https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification).

- We should try to make any SDK feature also available in the CLI.  Is this reasonable? Any thoughts on what a good API might look like?
- Finally, I think this may merit a brief section in the docs beyond the provided docstring.  What do you think about adding a brief section to https://docs.ray.io/en/latest/cluster/running-applications/job-submission/sdk.html?

architkulkarni commented on dashboard/modules/dashboard_sdk.py:
```diff
@@ -221,6 +224,20 @@ def __init__(
+                cafile, capath = None, self._verify
+            else:
+                cafile, capath = self._verify, None
+            self._ssl_context = ssl.create_default_context(cafile=cafile, capath=capath)
```
Oh, looks like `ssl` is in the standard library.  Then we should get rid of the conditional import at the top of the file, right?

shrekris-anyscale reviewed (approved):
Great work! Overall, it looks good to me, pending @architkulkarni's comments. Just had a clarifying question.

joncarter1 commented:
@architkulkarni 

> This looks good to me so far, and the new API makes sense because it matches the API in the `requests` library (https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification).
> 
>     * We should try to make any SDK feature also available in the CLI.  Is this reasonable? Any thoughts on what a good API might look like?

My concern with introducing explicit CLI support is: Does this also mean other client functionality e.g. metadata, cookies and HTTP headers, should also be supported?

A few thoughts on this (if yes to the above):
1. Since they are all used in all underlying client requests, they'd presumably all need CLI flags for all the `ray job` CLI commands for full CLI support. It should be possible to do this fairly cleanly using `click`.
2. How best to pass the others as CLI args? They're all dicts so either a JSON-serialized string or a path to a YAML file as done with the `runtime-env` seems reasonable.
3. CLI flag names? I don't see any problems with using the underlying variable names e.g. `--headers`, `--metadata`. Should there also be separate flags when using JSON-serialized versions like with `runtime-env`? e.g. `--metadata-json`. I'd be in favour of a single flag that could use either.

>     * Finally, I think this may merit a brief section in the docs beyond the provided docstring.  What do you think about adding a brief section to https://docs.ray.io/en/latest/cluster/running-applications/job-submission/sdk.html?

Was on the fence as to whether this merited more in the docs. Happy to add a brief section!

architkulkarni commented:
Re: docs, I'm also a little bit on the fence, but I think let's just add a brief section!

For the CLI issue, we have an work in progress PR to add metadata here https://github.com/ray-project/ray/pull/34586 where some of the JSON discussion also happened.

For `verify`, since it's just a bool or str, I wonder if we can ignore the JSON question for now and just do
- no arg provided -> `verify = True` (default) 
- `--verify` -> `verify = True` (weird but we can still support it)
- `--verify=true` -> `verify = True`
- `--verify=false` -> `verify = False`
- `--verify=<existing dir or file>` -> `verify` is the path

One problem here is how to handle a local certificate file or directory named `true` or `false`. Maybe we can pick one to prioritize in this case and print a warning if we detect this.

We can also leave this as a followup issue or PR to keep the current PR small.

joncarter1 commented:
Hey @architkulkarni,

I have addressed all suggestions and added a ``--verify`` flag to the CLI 😄

### Flag behaviour
The behaviour of the CLI flag is as follows:
```
  no arg provided -> verify = True (default)
  --verify={true,1} -> verify = True (bool)
  --verify={false,0} -> verify = False (bool)
  --verify=<path> -> verify = path (str)
```
I couldn't see an obvious way to make `--verify` also work as a flag without an argument.

### Bool / path ambiguity
As for handling files or directories named `true` or `false`: Since I've given the boolean priority, this should only be relevant to users trying to use custom certificate paths who presumably must have read the docs or CLI help to know about this feature.

I hope it should be clear, intuitively and from the CLI docs, to such a user why trying to use e.g. `--verify true` with a local path called `true` would fail. Turning a blind eye to (questionable!) OS-level decisions, if a cert location called (`true`,`false`,`0` or `1`) was somehow unavoidable it could presumably still be specified with an absolute path i.e. `/path/to/true`.

### CLI design
I have added the CLI flag using a new `@add_common_job_options` decorator which I have moved into a new `cli_utils.py` file. This should help with future CLI support for headers and cookies, reducing bloat/repeat code in the `cli.py` file and ensuring consistency across the `ray job` commands.

architkulkarni commented on dashboard/modules/job/cli_utils.py:
```diff
@@ -0,0 +1,46 @@
+    name = "BOOL | TEXT"
+
+    def convert(self, value, param, ctx):
+        if isinstance(value, bool):
```
Out of curiosity how could a bool (as opposed to a string "true") appear here? Isn't everything passed in the command line a string?

architkulkarni replied:
Answered in the click doc

> Values from user input or the command line will be strings, but default values and Python arguments may already be the correct type. The custom type should check at the top if the value is already valid and pass it through to support those cases.

architkulkarni commented on doc/source/cluster/running-applications/job-submission/sdk.rst:
```diff
@@ -223,4 +223,23 @@ To be precise, the environment variable ``CUDA_VISIBLE_DEVICES`` will not be set
+Additional client connection options, such as custom HTTP headers and cookies, can be passed to the ``JobSubmissionClient`` class. 
+A full list of options can be found in the :ref:`API Reference <ray-job-submission-sdk-ref>`.
+
+TLS Verification
+~~~~~~~~~
```
```suggestion
TLS Verification
~~~~~~~~~~~~~~~~
```
I think the linter might complain otherwise.

architkulkarni replied:
Oh, looks like it doesn't complain about this anymore!

architkulkarni reviewed (approved):
Looks great!

architkulkarni commented:
Test failures:
- Windows test_object_store_metrics unrelated
- `stablediffusion_batch_prediction` doc test unrelated
- `learning_tests_stateless_cartpole_r2d2` unrelated
- `test_dask_scheduler_client_mode` unrelated
